### PR TITLE
Add `libdispatch` package

### LINKED ISSUE #25678: Package request: libdispatch
### Package name

libdispatch

### Package version

Newest

### Package website

https://github.com/apple/swift-corelibs-libdispatch

### Package availability

https://github.com/apple/swift-corelibs-libdispatch
https://repology.org/project/libdispatch/information

### Additional comments

CMake is used as its build system. All three major OSes are supported.

### Package is not available

- [X] The package is not available on conda-forge.

### No previous issues or open PRs

- [X] No previous issue exists and no PR has been opened.

### PR DESCRIPTION
**Pull request by chawyehsu**
Close #25678 

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] ~~If static libraries are linked in, the license of the static library is packaged.~~
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/libdispatch/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  skip: true  # [osx]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - clangxx
+    - lld  # [win]
+    # It seems libdispatch cannot be built with glibc<2.19 because of: https://bugzilla.redhat.com/show_bug.cgi?id=1009623
+    # I have no idea how Apple build Swift (libdispatch is the dep) for CentOS 7 (glibc 2.17), its CI pipline is a blackbox
+    - sysroot_linux-64 2.28  # [linux64]
```
Comment (chawyehsu) [inline recipes/libdispatch/meta.yaml:22]: HELP WANTED: Either I find a way to mitigate this or I have to disable build for linux...

**Comment by chawyehsu** [comment]
@conda-forge/help-c-cpp, ready for review!

**Comment by chawyehsu** [comment]
@conda-forge-admin, please ping conda-forge/help-c-cpp

File: recipes/libdispatch/build.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+set -ex
+
+cmake ${CMAKE_ARGS} \
+    -G Ninja \
+    -B build \
+    -D BUILD_SHARED_LIBS=YES \
+    -D CMAKE_BUILD_TYPE=Release \
+    -D CMAKE_C_COMPILER=clang \
+    -D CMAKE_CXX_COMPILER=clang++ \
```
Comment (xhochy) [inline recipes/libdispatch/build.sh:11]: This should be compilers that contain a host triple. Please add the following to conda_build_config.yaml
```suggestion
c_compiler:  # [linux]
 - clang  # [linux]
cxx_compiler:  # [linux]
 - clang  # [linux]
```

File: recipes/libdispatch/bld.bat [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    -B build ^
+    -D BUILD_SHARED_LIBS=YES ^
+    -D CMAKE_BUILD_TYPE=Release ^
+    -D CMAKE_C_COMPILER=clang-cl ^
+    -D CMAKE_CXX_COMPILER=clang-cl ^
+    -D CMAKE_MT=mt ^
+    -D CMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+    -D CMAKE_PREFIX_PATH=%LIBRARY_PREFIX% ^
+    -D CMAKE_SYSTEM_NAME=Windows ^
+    -D CMAKE_SYSTEM_PROCESSOR=x86_64 ^
```
Comment (xhochy) [inline recipes/libdispatch/bld.bat:14]: This should be set via environment variables already and should be omitted in preparation for win-arm64
```suggestion
```

File: recipes/libdispatch/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - clangxx
+    - lld  # [win]
+    - sysroot_linux-64 2.17  # [linux64]
+    - cmake
+    - ninja
+  run_constrained:
+    - __glibc >=2.17  # [linux]
```
Comment (xhochy) [inline recipes/libdispatch/meta.yaml:26]: This is already set by the sysroot
```suggestion
```

File: recipes/libdispatch/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  version: {{ version }}
+
+source:
+  url: https://github.com/apple/swift-corelibs-libdispatch/archive/refs/tags/swift-{{ version }}-RELEASE.tar.gz
+  sha256: 16e088cf12654d22658879710b9694a6fad1c94d5e5d0c597741b71fbcb3e034
+  patches:
+    - 0001-glibc-unistd-wrapper.patch  # [linux]
+
+build:
+  number: 0
```
Comment (xhochy) [inline recipes/libdispatch/meta.yaml:14]: This is a shared library so it needs a run_exports.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/libdispatch/build.sh (hunk 2 above)
Comment (chawyehsu) [reply recipes/libdispatch/build.sh]: Changes have been applied as per the code review, however, it seems the compiler supported in the CI is forced to gcc and becuase of this failure happens on linux CI after clang is specified.

**Review by xhochy — APPROVED** [review]

**Pull request merged**